validator/client/runner.go

```diff
@@ -24,123 +24,197 @@ import (
 // Time to wait before trying to reconnect with beacon node.
 var backOffPeriod = 10 * time.Second
 
-// Run the main validator routine. This routine exits if the context is
-// canceled.
+// run is the main validator routine. It runs indefinitely until the context is canceled.
 //
-// Order of operations:
-// 1 - Initialize validator data
-// 2 - Wait for validator activation
-// 3 - Wait for the next slot start
-// 4 - Update assignments
-// 5 - Determine role at current slot
-// 6 - Perform assigned role, if any
+// Main responsibilities of the routine:
+// a) fetch and perform validator roles every slot,
+// b) periodically check if the beacon node is healthy and perform a switchover if possible
+// c) push proposer settings when appropriate
+// d) process events
+// e) act on account changes
 func run(ctx context.Context, v iface.Validator) {
 	cleanup := v.Done
 	defer cleanup()
 
-	headSlot, err := initializeValidatorAndGetHeadSlot(ctx, v)
-	if err != nil {
-		return // Exit if context is canceled.
+	if err := initializeValidator(ctx, v); err != nil {
+		log.WithError(err).Error("Could not initialize validator")
+		return
 	}
-	if err := v.UpdateDuties(ctx, headSlot); err != nil {
-		handleAssignmentError(err, headSlot)
+	headSlot, err := v.CanonicalHeadSlot(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get canonical head slot")
+		return
 	}
-	eventsChan := make(chan *event.Event, 1)
-	healthTracker := v.HealthTracker()
-	runHealthCheckRoutine(ctx, v, eventsChan)
 
-	accountsChangedChan := make(chan [][fieldparams.BLSPubkeyLength]byte, 1)
 	km, err := v.Keymanager()
 	if err != nil {
-		log.WithError(err).Fatal("Could not get keymanager")
+		log.WithError(err).Error("Could not get keymanager")
+		return
 	}
+	accountsChangedChan := make(chan [][fieldparams.BLSPubkeyLength]byte, 1)
 	sub := km.SubscribeAccountChanges(accountsChangedChan)
-	// check if proposer settings is still nil
-	// Set properties on the beacon node like the fee recipient for validators that are being used & active.
+
 	if v.ProposerSettings() == nil {
 		log.Warn("Validator client started without proposer settings such as fee recipient" +
 			" and will continue to use settings provided in the beacon node.")
 	}
-	deadline := time.Now().Add(5 * time.Minute)
-	if err := v.PushProposerSettings(ctx, km, headSlot, deadline); err != nil {
-		if errors.Is(err, ErrBuilderValidatorRegistration) {
-			log.WithError(err).Warn("Push proposer settings error")
-		} else {
-			log.WithError(err).Fatal("Failed to update proposer settings") // allow fatal. skipcq
+
+	pushProposerSettingsChan := make(chan struct{}, 1)
+	go func() {
```

Comment:
I wonder if breaking out some of these go routines into other functions would make this cleaner to review... there's a lot going on with this file and a lot of variables declared